README.md
###　　MicroPHP是一个免费、开源、敏捷的MVC模式的PHP开发框架。它易拓展而不失灵活，可以轻松对应大中小型项目。MicroPHP没有大而全的功能，只给你最需要的底层功能，但是给你无限的灵活拓展能力。其它框架都有的基本特性这里就不一一罗列。 

# 文件说明:    
    1.MicroPHP.php是整个框架核心程序  

    2.index.php是入口文件同时也是配置文件  

    3.MicroPHP.min.php是压缩版本，建议生产环境替代MicroPHP.php。  

    4.plugin.php是插件模式要被包含的文件，插件模式要保证plugin.php和MicroPHP.min.php在一起，  
      或者修改plugin.php最后的包含MicroPHP.min.php的路径。  
    
    5.application是一个简单的项目结构   

官方首页：http://microphp.us/    
# composer方式安装最新版microphp:    
    1.新建一个空文件夹，比如 microphp  
    
    2.在microphp文件夹里面执行命令：  
      composer create-project snail/microphp  ./ dev-master  
      dev-master是最新版，也可以用版本号，比如：v2.3.1  

    3.在index.php里面的MpRouter::setConfig($system)的下面  
      执行包含：require dirname(__FILE__) . '/vendor/autoload.php';  

# 说一下MicroPHP的特性：  

    1. 整个框架核心就一个文件MicroPHP.php（压缩后的文件MicroPHP.min.php，140KB左右）。   

    2. 入口文件index.php也是整个项目的配置文件，一个入口文件就是一个项目的入口，可以自定义多个入口文件达到不同项目共存的目的，项目之间随意共享类库、帮助文件等目录，框架核心文件采用零侵入式编写，整个框架的运行不依赖任何外部变量，   
       初始化的时候只要通过MpRouter::setConfig($system)注入配置，然后MpRouter::loadClass()执行即可。   
    
    3. 独特的插件模式。  
    
    4. 项目控制器、模型、视图、类库等各种文件夹完全自定义，用到什么就建立什么文件夹，彻底摆脱一堆无用的一大堆文件夹的困扰。  
    
    5. 类库采用懒加载模式，只要按着规定命名规则，然后把你的类库文件扔到类库文件夹，在控制器或者模型中直接new 类库名()即可，系统会自动加载相应的类库文件。    
    
    6. 没有视图就不能指定视图数据？文件可以include共享那么视图为什么不能共享数据呢？在MicroPHP里面你不再有这个困挠，MicroPHP可以在  
       $this->view_vars数组里面存放你的任何想在视图里面使用的全局数据。 特别是网站头部导航用户数据全站每个页面都用到，那么可以自定义个控制器父类，然后和在父类构造方法里面初始化这个用户数据，放到$this->view_vars里面，那么所有的控制器再也不用重复的去取用户数据传给视图。   
    
    7. 灵活的session托管,支持的管理类型：mongodb,mysql,memcache,redis.当然也可以用系统默认的管理方式.session托管是可选的,根据项目具体情况择优选用即可。session托管的好处很多，比如：支持分布式、精确控制session过期时间，等等。    
    
    8. 灵活的缓存机制。   
        (1).可用的方式缓存驱动有：auto,apc,sqlite,files,memcached,redis,wincache,xcache,memcache。   
            auto自动模式寻找的顺序是 : apc,sqlite,files,memcached,redis,wincache,xcache,memcache。   
        (2).缓存配置有个第二驱动机制，比如：当你现在在代码中使用的是memcached, apc等等，然后你的代码转移到了一个新的服务器   
            而且不支持memcached 或 apc这时候怎么办呢？设置第二驱动即可，当你设置的驱动不支持的时候，系统就使用第二驱动。   

更多详细信息,请移步官网:http://microphp.us/

CHANGELOG.txt
#############################ver 2.3.2[2015-05-15]#############################
修复了：
1.apache伪静态支持问题,新版apache在.htaccess中开启了伪静态在之后，
  $_SERVER里面的PATH_INFO变量变了名字为：REDIRECT_PATH_INFO

#############################ver 2.3.1[2015-03-13]#############################
优化了：
1.分页条显示，“转到”默认没有下划线
2.其它一些代码优化

增加了：
2.增加了composer支持，麻麻再也不用担心我的类库乱七八糟了。

修复了：
1.get_post($key,$default),post_get($key,$default),
  当$key存在值，如果传递了默认值$default将返回空的bug。

题外话，这次更新距离上次时间很久了，并不是停止开发了。年前入职了新浪后，学到了很多思想，审视现在框架有很多值得革命性优化的地方。所以启动了另外一个新项目，开发阶段，透漏一下新版借鉴了很多思想，解决了其它框架的短板，并把优点强化，保持microphp一贯的灵活自由，更小，更方便，更优雅，不久将会和大家见面。microphp会同步保持活跃更新的大家放心使用，新版让大家有了更多的选择。另外，很多人肯定认为这是重复造轮子，但是我认为有创新的造轮子是必要的，不然互联网甚至任何行业都会停止在前人的脚步上。

#############################ver 2.3.0[2014-10-18]#############################
优化了：
1.控制器类、模型类、Loader、输入类、规则类使用标准名称:MpController、MpModel、MpLoader、MpInput、MpRule
  以前的WoniuController、WoniuModel、WoniuInput、WoniuRule可以继续使用，兼容以前版本。
2.路由器获取参数的规则。

升级提示：
1.修改入口文件里面最后两行
  WoniuRouter::setConfig($system);
  WoniuRouter::loadClass();
  修改为：
  MpRouter::setConfig($system);
  MpRouter::loadClass();
  即可。
2.如果使用了插件模式，修改插件入口文件plugin.php里面最后一行
  WoniuRouter::setConfig($system);
  修改为：
  MpRouter::setConfig($system);
  即可。
3.如果自定义了loader，ver2.2.14以前是自定义了loader必须继承WoniuLoader,现在修改为继承MpLoader即可。
4.如果自己使用了WoniuRouter::setConfig($system);WoniuRouter::loadClass();
  修改为MpRouter::setConfig($system);MpRouter::loadClass();即可

#############################ver 2.2.13[2014-10-07]#############################
优化了：
1.精简了托管session的几个类
2.优化了session管理机制，对于session通用配置$system['session_handle']['common']，无论系统是否托管session，都会起作用。
3.精简了数据库驱动

增加了：
1.memcached存储session，对应的系统配置是$system['session_handle']['memcached']
2.为$this->input->xxx增加了一批快捷xxx的函数,参数一致，具体参数可以参考手册“输入”部分，xxx具体如下：
    args()                    对应：$this->input->parameters()
    xss_clean()               对应：$this->input->xss_clean()
    server()                  对应：$this->input->server()
    session()                 对应：$this->input->session()
    get()                     对应：$this->input->get()
    post()                    对应：$this->input->post()
    cookie()                  对应：$this->input->cookie()
    cookie_raw()              对应：$this->input->cookieRaw()
    get_post()                对应：$this->input->get_post()
    post_get()                对应：$this->input->post_get()
    get_int()                 对应：$this->input->get_int()
    post_int()                对应：$this->input->post_int()
    get_post_int()            对应：$this->input->get_post_int()
    post_get_int()            对应：$this->input->post_get_int()
    get_time()                对应：$this->input->get_time() 
    post_time()               对应：$this->input->post_time() 
    get_post_time()           对应：$this->input->get_post_time()
    post_get_time()           对应：$this->input->post_get_time()
    get_datetime()            对应：$this->input->get_datetime()
    post_datetime()           对应：$this->input->post_datetime()
    get_post_datetime()       对应：$this->input->get_post_datetime()
    post_get_datetime()       对应：$this->input->post_get_datetime()
    get_rule()                对应：$this->input->get_rule()
    post_rule()               对应：$this->input->post_rule()
    get_post_rule()           对应：$this->input->get_post_rule()
    post_get_rule()           对应：$this->input->post_get_rule()
3.增加了下面快捷函数,参数一致，具体参数可以参考手册“输出”部分
    set_cookie()           对应：$this->input->setCookie()
    set_cookie_raw()       对应：$this->input->setCookieRaw()

修复了：
1.$this->input->xxx()默认值有时不起作用的问题。

#############################ver 2.2.12[2014-09-01]#############################
增加了：
1.$this->lib($file_name, $alias_name = null,$new=true),增加了第三个参数$new，默认true。
  默认情况下，$this->lib()加载类库的时候，会new一个类的对象。
  有时候我们只是为了包含进来类库文件，或者类是不可实例化的，那么我们只要指定$new=false即可。
  然后$this->lib()加载类库的时，只是包含文件，并不会new一个类的对象，此时xxx类对应的$this->lib->xxx不再可用，是null。
2.系统配置中，类库自动加载的时候，可以指定new参数。
  $system['library_file_autoload'] = array();
  自动加载的library文件,比如array($item); 
  $item是library文件名或者"配置数组"，不包含后缀。
  比如: ImageTool 或者配置数组array('ImageTool'=>'image'),或者配置数组array('ImageTool'=>'image','new'=>fasle)
  配置数组的作用是为长的类库名用别名代替.
3.$this->helper($file_name, $is_config = true),增加了第二个参数$is_config，默认true。
  默认情况下，$this->helper()加载helper文件的时候，会把文件里面的全局变量放到self::$config里面。
  有时候我们只想要包含helper文件，不对里面的变量进行处理。那么可以通过指定第二个参数$is_config=false即可。
4.系统配置中，helper自动加载的时候，可以指定$is_config参数。
  $system['helper_file_autoload] = array();
  自动加载的helper文件,比如array($item); 
  $item是helper文件名或者"配置数组"，不包含后缀。
  比如: html 或者配置数组array('html'=>false),这里的数组key是helper文件，值就是$is_config参数，是true或者false。

修复了：
1.sqlite_escape_string在PHP5.4及以上的版本中被废弃了，使用SQLite3::escapeString()代替。
2.表模型的search方法报错的问题。

优化了：
1.精简了缓存类
2.$this->page()产生的分页条，当页数大于2页时再显示，这样避免了出现尴尬的一页的时候，还有下一页。
3.优化了对路由字符串的解析，能够支持get表单通过name指定要访问的控制器方法。
4.移除了$system['folder_auto_init']系统配置，下载的代码中已经包含了完整的文件夹结构，
  这个自动建立文件夹功能有些鸡肋，为了提高性能移除掉。
5.优化了xss过滤方法：$this->input->xss_clean()。

升级提示：
无

#############################ver 2.2.11[2014-07-28]#############################
优化了：
1.缓存启动，自动模式搜索顺序修改为：apc,files,sqlite,memcached,redis,wincache,xcache,memcache
修复了:
1.使用mysql存储session的时候，session不会自动延期的问题。
2.使用mysqli数据库驱动的时候，解决了报Deprecated: mysql_escape_string()的问题，5.3以后不再使用该函数。

#############################ver 2.2.10[2014-06-24]#############################
增加了：
1.增加了$this->input->route_query()方法，返回的是系统最终使用的路由字符串.
  该方法在自定义了路由之后，方面查看自己的路由规则替换的结果是否正常。
2.$this->input->xss_clean($var)，$var支持数组了，一维或者多维数组都可以。
修复了:
1.缓存驱动为files和sqlite时，设置过期时间为0，让缓存数据不过期，无效的问题。

#############################ver 2.2.9[2014-06-01]#############################
增加了：
1.增加了表单验证exists[table.col,col1:val1,...]规则。
    规则说明：
       (1).如果表单元素的值在指定数据表的字段中不存在则返回false，如果存在返回true
       (2).比如exists[cat.cid]，那么验证类会去查找cat表中cid字段有没有与表单元素一样的值
       (3).cat.cid，cat是表名称（不要带前缀）cid是字段名称。
         cat.cid后面还可以指定附加的where条件。
         比如：exists[users.uname,user_id:2,...] 可以多个条件，逗号分割。
         上面的规则生成的where条件就是array('uname'=>$value,'user_id'=>2,....)
2.cookie键前缀配置，$system['cookie_key_prefix'] = '';

index.php
<?php
/*
 * Copyright 2015 狂奔的蜗牛.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * MicroPHP
 *
 * An open source application development framework for PHP 5.2.0 or newer
 *
 * @package       MicroPHP
 * @author        狂奔的蜗牛
 * @email         672308444@163.com
 * @copyright     Copyright (c) 2013 - 2015, 狂奔的蜗牛, Inc.
 * @link          http://git.oschina.net/snail/microphp
 * @since         Version 2.3.2
 * @createdtime   2015-05-15 15:49:19
 */
 

define('IN_WONIU_APP', TRUE);
define('WDS', DIRECTORY_SEPARATOR);
/**
 * --------------------系统配置-------------------------
 */
/**
 * 如果开启了URL Rewrite 功能，请在这里设置为TRUE，没有开启请设置为FALSE
 * 该配置会影响url()函数生成的链接
 */
$system['url_rewrite'] = FALSE;
/**
 * 程序文件夹路径名称，也就是所有的程序文件比如控制器文件夹，
 * 模型文件夹，视图文件夹等所在的文件夹名称。
 */
$system['application_folder'] = dirname(__FILE__) . '/' . 'application';
/**
 * 存放控制器文件的文件夹路径名称
 */
$system['controller_folder'] = $system['application_folder'] . '/controllers';
/**
 * 存放模型文件的文件夹路径名称,支持数组
 */
$system['model_folder'] = $system['application_folder'] . '/models';
/**
 * 存放视图文件的文件夹路径名称,支持数组
 */
$system['view_folder'] = $system['application_folder'] . '/views';
/**
 * 存放类库文件的文件夹路径名称,存放在该文件夹的类库中的类会自动加载,支持数组
 */
$system['library_folder'] = $system['application_folder'] . '/library';
/**
 * 存放函数文件的文件夹路径名称,支持数组
 */
$system['helper_folder'] = $system['application_folder'] . '/helper';
/**
 * table()方法缓存表字段信息的文件夹路径名称
 */
$system['table_cache_folder'] = $system['application_folder'] . '/cache';
/**
 * 存放HMVC模块的文件夹路径名称
 */
$system['hmvc_folder'] = $system['application_folder'] . '/modules';
/**
 * 注册HMVC模块，这里填写模块名称关联数组,键是url中的模块别名，值是模块文件夹名称
 */
$system['hmvc_modules'] = array('demo' => 'hmvc_demo');
/**
 * 404错误文件的路径,该文件会在系统找不到相关内容时显示,
 * 文件里面可以使用$msg变量获取出错提示内容
 */
$system['error_page_404'] = 'application/error/error_404.php';
/**
 * 系统错误文件的路径,该文件会在发生Fatal错误和Exeption时显示,
 * 文件里面可以使用$msg变量获取出错提示内容
 */
$system['error_page_50x'] = 'application/error/error_50x.php';
/**
 * 数据库错误文件的路径,该文件会在发生数据库错误时显示,
 * 文件里面可以使用$msg变量获取出错提示内容
 */
$system['error_page_db'] = 'application/error/error_db.php';
/**
 * $this->message()方法默认使用的视图，该视图会在第4个参数为null时使用。
 * 视图里面可以使用的有三个变量：$msg提示内容，$url跳转的url，$time停留时间。
 * 这里需要填写的是视图名称，不带视图后缀。没有就留空。
 */
$system['message_page_view'] = '';
/**
 * 默认控制器文件名称,不包含后缀,支持子文件夹,比如home.welcome,
 * 就是控制器文件夹下面的home文件夹里面welcome.php(假定后缀是.php)
 */
$system['default_controller'] = 'welcome';
/**
 * 默认控制器方法名称,不要带前缀
 */
$system['default_controller_method'] = 'index';
/**
 * 控制器方法名称前缀
 */
$system['controller_method_prefix'] = 'do';
/**
 * 控制器文件名称后缀,比如.php或者.controller.php
 */
$system['controller_file_subfix'] = '.php';
/**
 * 模型文件名称后缀,比如.model.php
 */
$system['model_file_subfix'] = '.model.php';
/**
 * 视图文件名称后缀,比如.view.php
 */
$system['view_file_subfix'] = '.view.php';
/**
 * 类库文件名称后缀,比如.class.php
 */
$system['library_file_subfix'] = '.class.php';
/**
 * 函数文件名称后缀,比如.php
 */
$system['helper_file_subfix'] = '.php';
/**
 * $this->input->setCookie()和$this->input->cookie()
 * 设置和获取cookie的时候key使用的前缀
 * 使用前缀的目的：
 * 避免主域名和子域名设置的cookie冲突
 */
$system['cookie_key_prefix'] = '';
/**
 * 自定义Loader，用于拓展框架核心功能,
 * Loader是控制器和模型都继承的一个类，大部分核心功能都在loader中完成。
 * 这里是自定义Loader类文件的完整路径
 * 自定义Loader文件名称和类名称必须是：
 * 文件名称：类名.class.php
 * 比如：MyLoader.class.php，文件里面的类名就是:MyLoader
 * 注意：
 * 1.自定义Loader必须继承MpLoader。
 * 2.一个最简单的Loader示意：(假设文件名称是：MyLoader.class.php)
 * class MyLoader extends MpLoader {
 *      public function __construct() {
 *          parent::__construct();
 *      }
 *  }
 * 3.如果无需自定义Loader，留空即可。
 * 4.自定义Loader在框架核心文件被包含时生效，此后修改$system['my_loader']无效。
 */
$system['my_loader'] = '';
/**
 * 自动加载的helper文件,比如:array($item);
 * $item是helper文件名,不包含后缀,比如: html 等.
 */
$system['helper_file_autoload'] = array();
/**
 * 自动加载的library文件,比如array($item);
 * $item是library文件名或者"配置数组",不包含后缀,
 * 比如: ImageTool 或者配置数组array('ImageTool'=>'image'),或者配置数组array('ImageTool'=>'image','new'=>fasle)
 * 配置数组的作用是为长的类库名用别名代替.
 */
$system['library_file_autoload'] = array();
/**
 * 自动加载的model,比如array($item);
 * $item是model文件名或者"配置数组",不包含后缀,
 * 比如: UserModel 或者配置数组 array('UserModel'=>'user')
 * 配置数组的作用是为长的model名用别名代替.
 */
$system['models_file_autoload'] = array();
/**
 * 控制器方法名称是否首字母大写,默认true
 */
$system['controller_method_ucfirst'] = TRUE;
/**
 * 是否自动连接数据库,默认FALSE
 */
$system['autoload_db'] = FALSE;
/**
 * 是否开启调试模式
 * true：显示错误信息,
 * false：所有错误将不显示
 */
$system['debug'] = TRUE;
/**
 * 是否接管错误信息显示
 * true：所有错误信息将由系统格式化输出
 * false：所有错误信息将原样输出
 */
$system['error_manage'] = FALSE;
/**
 * 是否开启错误日志记录
 * true：开启，如果开启了，系统将接管错误信息输出，忽略system['error_manage']和$system['db']['xxx']['db_debug']，
 *       同时务必设置自己的错误日志记录处理方法
 * false：关闭
 * 提示：
 * 数据库错误信息是否显示是由：$system['debug']和db_debug（$system['db']['xxx']['db_debug']）控制的。
 * 只用都为TRUE时才会显示。
 */
$system['log_error'] = FALSE;
/* * --------------------------------错误日志记录处理配置-----------------------
 * 错误日志记录处理方法，可以是一个“函数名称”或是“类的静态方法”用数组方式array('class_name'=>'method_name')。
 * 提示：
 * 1.如果是类，把类按着类库的命名方式命名，然后放到类库目录即可;
 * 2.如果是函数，把函数放到一个helper文件里面，然后在$system['helper_file_autoload']自动加载的helper文件里面填写上这个helper文件即可。
 * 3.留空则不处理。
 * 4.系统会传递给error、exception处理方法5个参数：（$errno, $errstr, $errfile, $errline,$strace）
 * 参数说明：
 * $errno：错误级别，就是PHP里面的E_NOTICE之类的静态变量，错误级别和具体含义对应关系如下，键是代码，值是代码含义。
 *         array('0'=>'EXCEPTION',//异常信息
 *               '1' => 'ERROR',//致命的运行时错误。这类错误一般是不可恢复的情况，例如内存分配导致的问题。后果是导致脚本终止不再继续运行。
 *               '2' => 'WARNING', //运行时警告 (非致命错误)。仅给出提示信息，但是脚本不会终止运行。
 *               '4' => 'PARSE', //编译时语法解析错误。解析错误仅仅由分析器产生。
 *               '8' => 'NOTICE', //运行时通知。表示脚本遇到可能会表现为错误的情况，但是在可以正常运行的脚本里面也可能会有类似的通知。
 *               '16' => 'CORE_ERROR', //在PHP初始化启动过程中发生的致命错误。该错误类似 E_ERROR，但是是由PHP引擎核心产生的。
 *               '32' => 'CORE_WARNING',//PHP初始化启动过程中发生的警告 (非致命错误) 。类似 E_WARNING，但是是由PHP引擎核心产生的。
 *               '64' => 'COMPILE_ERROR', //致命编译时错误。类似E_ERROR, 但是是由Zend脚本引擎产生的。
 *               '128' => 'COMPILE_WARNING', //编译时警告 (非致命错误)。类似 E_WARNING，但是是由Zend脚本引擎产生的。
 *               '256' => 'USER_ERROR', //用户产生的错误信息。类似 E_ERROR, 但是是由用户自己在代码中使用PHP函数 trigger_error()来产生的。
 *               '512' => 'USER_WARNING', //用户产生的警告信息。类似 E_WARNING, 但是是由用户自己在代码中使用PHP函数 trigger_error()来产生的。
 *               '1024' => 'USER_NOTICE',//用户产生的通知信息。类似 E_NOTICE, 但是是由用户自己在代码中使用PHP函数 trigger_error()来产生的。
 *               '2048' => 'STRICT', //启用 PHP 对代码的修改建议，以确保代码具有最佳的互操作性和向前兼容性。
 *               '4096' => 'RECOVERABLE_ERROR'//可被捕捉的致命错误。 它表示发生了一个可能非常危险的错误，但是还没有导致PHP引擎处于不稳定的状态。 如果该错误没有被用户自定义句柄捕获 (参见 set_error_handler())，将成为一个 E_ERROR　从而脚本会终止运行。
 *               '8192' => 'DEPRECATED', //（php5.3）运行时通知。启用后将会对在未来版本中可能无法正常工作的代码给出警告。
 *               '16384' => 'USER_DEPRECATED', //（php5.3）用户产少的警告信息。 类似 E_DEPRECATED, 但是是由用户自己在代码中使用PHP函数 trigger_error()来产生的。
 *         );
 *         可以通过判断错误级别，然后有针对性的处理。一般我们需要处理的就是致命错误（0，1，4）和一般错误（2，8，2048，8192）.
 * $errstr：具体的错误信息
 * $errfile：出错的文件完整路径
 * $errline：出错的行号
 * $strace： 调用堆栈信息
 * 系统会传递给db_error处理方法2个参数：（$errmsg,$strace）
 * 参数说明：
 * $errmsg：具体的数据库错误信息
 * $strace：调用堆栈信息
 * 错误控制类参考：
 * /blob/development/tests/app/library/ErrorHandle.class.php
 */
$system['log_error_handle'] = array(
	'error' => '', //array('ErrorHandle' => 'error_handle'),
	'exception' => '', //array('ErrorHandle' => 'exception_handle'),
	'db_error' => '', //array('ErrorHandle' => 'db_error_handle')
);
/**
 * 默认时区,PRC是中国
 */
$system['default_timezone'] = 'PRC';
/**
 * ---------------------------自定义URL路由规则------------------------
 * 比如：
 *   (1).http://localhost/index.php?welcome.index
 *   (2).http://localhost/index.php/welcome.index
 *   (3).http://localhost/index.php?news/welcome.index
 *   (4).http://localhost/index.php/news/welcome.index
 * 路由字符串是welcome.index(不包含最前面的?、/、模块名称)，路由规则都是针对“路由字符串”的。
 * 现在定义路由规则：
 *   $system['route']=array(
 *        "/^welcome\\/?(.*)$/u"=>'welcome.ajax/$1'
 *   );
 * 路由规则说明：
 *  1.路由规则是一个关联数组
 *  2.数组的key是匹配“路由字符串”的正则表达式，其实就是preg_match的第一个参数。
 *  3.数组的value是替换后的路由字符串
 *  4.系统使用的url路由就是最后替换后的路由字符串
 */
$system['route'] = array(
	//'|^([^/]+)/(.+)$|u' => '$1.$2',//index.php/home/index路由支持
	//'|^([^/]+)/([^/]+)/(.+)$|u' => '$1.$2.$3',//index.php/user/home/index路由支持
);
/**
 * ---------------------缓存配置-----------------------
 */
/**
 * 自定义缓存类文件的路径是$system['cache_drivers']的一个元素，
 * 可以有多个自定义缓存类。
 * 缓存类文件名称命名规范是：
 * 比如文件名是mycahe.php,那么文件mycahe.php
 * 里面的缓存类就是：class phpfastcache_mycahe{......}
 * mycahe.php的编写规范请参考：
 * /blob/development/modules/cache-drivers/drivers/example.php
 */
$system['cache_drivers'] = array();
/**
 * 缓存配置项
 */
$system['cache_config'] = array(
	/*
	 * 默认存储方式
	 * 可用的方式有：auto,apc,files,sqlite,memcached,redis,wincache,xcache,memcache
	 * auto自动模式寻找的顺序是 : apc,files,sqlite,memcached,redis,wincache,xcache,memcache
	 */
	"storage" => "auto",
	/*
	 * 默认缓存文件存储的路径
	 * 使用绝对全路径，比如： /home/username/cache
	 * 留空，系统自己选择
	 */
	"path" => $system['application_folder'] . "/cache", // 缓存文件存储默认路径,使用files、sqlite缓存的时候确保文件夹存在而且可写
	/*
	 * 第二驱动
	 * 比如：当你现在在代码中使用的是memcached, apc等等，然后你的代码转移到了一个新的服务器而且不支持memcached 或 apc
	 * 这时候怎么办呢？设置第二驱动即可，当你设置的驱动不支持的时候，系统就使用第二驱动。
	 * $key是你设置的驱动，当设置的“storage”=$key不可用时，就使用$key对应的$value驱动
	 */
	"fallback" => array(
		"memcache" => "files",
		"memcached" => "files",
		"redis" => "files",
		"wincache" => "files",
		"xcache" => "files",
		"apc" => "files",
		"sqlite" => "files",
	),
	/*
	 * Memcache服务器地址;
	 */
	"server" => array(
		array("127.0.0.1", 11211, 1),
	),
	/*
	 * Redis服务器地址;
	 */
	"redis" => array(
		'type' => 'tcp', //sock,tcp;连接类型，tcp：使用host port连接，sock：本地sock文件连接
		'prefix' => @$_SERVER['HTTP_HOST'], //key的前缀，便于管理查看，在set和get的时候会自动加上和去除前缀，无前缀请保持null
		'sock' => '', //sock的完整路径
		'host' => '127.0.0.1',
		'port' => 6379,
		'password' => NULL, //密码，如果没有,保持null
		'timeout' => 0, //0意味着没有超时限制，单位秒
		'retry' => 100, //连接失败后的重试时间间隔，单位毫秒
		'db' => 0, // 数据库序号，默认0, 参考 http://redis.io/commands/select
	),
);
/**
 * -----------------------SESSION管理配置---------------------------
 */
$system['session_handle'] = array(
	'handle' => '', //支持的管理类型：mongodb,mysql,memcache,memcached,redis。留空则不管理，使用默认
	'common' => array(//SESSION公共配置，无论是否使用handle托管session，common配置都会起作用。
		'autostart' => false, //是否自动session_start()
		'cookie_path' => '/',
		'cookie_domain' => empty($_SERVER['HTTP_HOST']) ? null : $_SERVER['HTTP_HOST'],
		'session_name' => 'MICROPHP',
		'lifetime' => 3600, // session lifetime in seconds
	),
	'mongodb' => array(
		'host' => '127.0.0.1',
		'port' => 27017,
		'user' => 'root',
		'password' => 'local',
		'database' => 'local', // name of MongoDB database
		'collection' => 'session', // name of MongoDB collection
		'persistent' => false, // persistent connection to DB?
		'persistentId' => 'MongoSession', // name of persistent connection
		'replicaSet' => false,
	),
	/**
	 * mysql表结构
	 *   CREATE TABLE `session_handler_table` (
	  `id` varchar(255) NOT NULL,
	  `data` mediumtext NOT NULL,
	  `timestamp` int(255) NOT NULL,
	  PRIMARY KEY (`id`),
	  UNIQUE KEY `id` (`id`,`timestamp`),
	  KEY `timestamp` (`timestamp`)
	  ) ENGINE=InnoDB DEFAULT CHARSET=utf8;
	 */
	'mysql' => array(
		'host' => '127.0.0.1',
		'port' => 3306,
		'user' => 'root',
		'password' => 'admin',
		'database' => 'test',
		'table' => 'session_handler_table',
	),
	/**
	 * memcached采用的是session.save_handler管理机制
	 * 需要php安装memcached拓展支持
	 */
	'memcached' => "127.0.0.1:11211",
	/**
	 * memcache采用的是session.save_handler管理机制
	 * 需要php安装memcache拓展支持
	 */
	'memcache' => "tcp://127.0.0.1:11211",
	/**
	 * redis采用的是session.save_handler管理机制
	 * 需要php安装redis拓展支持,你可以在https://github.com/nicolasff/phpredis 找到该拓展。
	 */
	'redis' => "tcp://127.0.0.1:6379",
);
/**
 * ------------------------数据库配置----------------------------
 */
/**
 * 默认使用的数据库组名称，名称就是下面的$system['db'][$key]里面的$key，
 * 可以自定义多个数据库组，然后根据不同的环境选择不同的组作为默认数据库连接信息
 */
$system['db']['active_group'] = 'default';
/**
 * dbdriver：可用的有mysql,mysqli,pdo,sqlite3,配置见下面
 */
/**
 * mysql数据库配置示例,如果用mysqli，把下面的dbdriver驱动由mysql改成mysqli即可
 */
$system['db']['default']['dbdriver'] = "mysql";
$system['db']['default']['hostname'] = '127.0.0.1';
$system['db']['default']['port'] = '3306';
$system['db']['default']['username'] = 'root';
$system['db']['default']['password'] = 'admin';
$system['db']['default']['database'] = 'test';
$system['db']['default']['dbprefix'] = '';
$system['db']['default']['pconnect'] = FALSE;
$system['db']['default']['db_debug'] = TRUE;
$system['db']['default']['char_set'] = 'utf8';
$system['db']['default']['dbcollat'] = 'utf8_general_ci';
$system['db']['default']['swap_pre'] = '';
$system['db']['default']['autoinit'] = TRUE;
$system['db']['default']['stricton'] = FALSE;
/*
 * PDO database config demo
 * 1.pdo sqlite3
 * */
/**
 * sqlite3数据库配置示例
 */
$system['db']['sqlite3']['dbdriver'] = "sqlite3";
$system['db']['sqlite3']['database'] = 'sqlite:d:/wwwroot/sdb.db';
$system['db']['sqlite3']['dbprefix'] = '';
$system['db']['sqlite3']['db_debug'] = TRUE;
$system['db']['sqlite3']['char_set'] = 'utf8';
$system['db']['sqlite3']['dbcollat'] = 'utf8_general_ci';
$system['db']['sqlite3']['swap_pre'] = '';
$system['db']['sqlite3']['autoinit'] = TRUE;
$system['db']['sqlite3']['stricton'] = FALSE;
/**
 * PDO mysql数据库配置示例，hostname 其实就是pdo的dsn部分，
 * 如果连接其它数据库按着pdo的dsn写法连接即可
 */
$system['db']['pdo_mysql']['dbdriver'] = "pdo";
$system['db']['pdo_mysql']['hostname'] = 'mysql:host=localhost;port=3306';
$system['db']['pdo_mysql']['username'] = 'root';
$system['db']['pdo_mysql']['password'] = 'admin';
$system['db']['pdo_mysql']['database'] = 'test';
$system['db']['pdo_mysql']['dbprefix'] = '';
$system['db']['pdo_mysql']['db_debug'] = TRUE;
$system['db']['pdo_mysql']['char_set'] = 'utf8';
$system['db']['pdo_mysql']['dbcollat'] = 'utf8_general_ci';
$system['db']['pdo_mysql']['swap_pre'] = '';
$system['db']['pdo_mysql']['autoinit'] = TRUE;
$system['db']['pdo_mysql']['stricton'] = FALSE;
/**
 * -------------------------数据库配置结束--------------------------
 */
/* End of file index.php */
include('MicroPHP.min.php');
MpRouter::setConfig($system);
MpRouter::loadClass();